### ISSUE TITLE
Feature request: lock bound nodes and relationships in pattern when deleting relationship, to set properties on bound nodes without duplication on concurrent writes

### ISSUE THREAD

**Issue by maiieul** [neo4j/neo4j#12819]
**Is your feature request related to a problem? Please describe.**

For better indexing, I'd like to increment a property on a node everytime a relationship is merged and decrement the same property everytime the relationship gets deleted.

Here is an example:

If I understand correctly, the following `MERGE` query effectively sets the properties ONLY ONCE on create, even on concurrent writes [by locking the bound nodes and the relationship](https://neo4j.com/developer/kb/understanding-how-merge-works/#_merge_acquires_locks_on_nodes_and_relationships_in_the_pattern_when_resulting_in_pattern_creation).

```
MATCH (p:Post{id:$postId})
MATCH (u:User{id:$userId})
MERGE (u)-[up:UPVOTED_POST]->(p)
ON CREATE SET p.numberOfUpvotes = p.numberOfUpvotes + 1
RETURN p, u
```

And to reproduce same behavior and decrement with `DELETE`, I've only found the following solution to work using [apoc.lock.nodes](https://neo4j.com/labs/apoc/4.1/overview/apoc.lock/apoc.lock.nodes/):

```
MATCH (p:Post{id:$postId})
MATCH (u:User{id:$userId})
CALL apoc.lock.nodes([p, u])
MATCH (u)-[up:UPVOTED_POST]->(p)
DELETE up
SET p.numberOfUpvotes = p.numberOfUpvotes - 1
RETURN p, u
```
Without apoc.lock.nodes, concurrent writes will decrement the p.numberOfUpvotes property twice, resulting in an unreliable way to count the upvotes. This is because on concurrent writes, the `MATCH (u)-[up:UPVOTED_POST]->(p)` clause will effectively match in each concurrent query, the `DELETE up` clause will effectively return an empty `up` object in each concurrent query (so it's not possible to use that in a reliable way to conditionnally set the numberOfUpvotes) and the `SET p.numberOfUpvotes = p.numberOfUpvotes - 1` clause will effectively decrement `p.numberOfUpvotes` twice. Thus, `CALL apoc.lock.nodes([p, u])` or another approach to lock bound nodes is needed.

**Describe the solution you'd like**

[Similar to `MERGE`](https://neo4j.com/developer/kb/understanding-how-merge-works/#_merge_acquires_locks_on_nodes_and_relationships_in_the_pattern_when_resulting_in_pattern_creation), When `DELETE` is used on a relationship, it would acquire locks on all bound nodes and relationships in the pattern (on `u`, `up` and `p` in my example queries) before deleting the relationship.

**Describe alternatives you've considered**

Before finding the `CALL apoc.lock.nodes([p, u])` solution, I thought about an `ON DELETE SET syntax`:
```
MATCH (p:Post{id:$postId})
MATCH (u:User{id:$userId})
OPTIONAL MATCH (u)-[up:UPVOTED_POST]->(p)
DELETE up
ON DELETE SET p.numberOfUpvotes = p.numberOfUpvotes - 1
RETURN p
```
The `ON DELETE` could be the syntax to add locks on bound nodes and relationships.

**Additional context**
Understanding the issue at play and finding the `CALL apoc.lock.nodes([p, u])` solution took me about 3 days and a bit of luck. If this feature can't be implemented for reasons I didn't consider, it would be nice to have the workaround documented somewhere, at least on stack-overflow.

**Comment by maiieul** [comment]
Oh and just like `MERGE`, [performing another match afther locking on bound elements would be necessary to avoid race conditions.](https://neo4j.com/developer/kb/understanding-how-merge-works/#_merge_acquires_locks_on_nodes_and_relationships_in_the_pattern_when_resulting_in_pattern_creation)

You can reproduce the decrementing issue by doing the following:

1. create user, post and merge relationship
```
CREATE (p:Post{id:"1", numberOfUpvotes: 0})
CREATE (u:User{id:"2"})
MERGE (u)-[:UPVOTED_POST]->(p)
SET p.numberOfUpvotes = p.numberOfUpvotes + 1
RETURN p, u
```
2. Run following query concurrently (you have 15 secs to run it at least twice :tongue: )
```
MATCH (p:Post{id:"1"})
MATCH (u:User{id:"2"})
MATCH (u)-[up:UPVOTED_POST]->(p)
CALL apoc.util.sleep(15000)
DELETE up
SET p.numberOfUpvotes = p.numberOfUpvotes - 1
RETURN p
```


- Neo4j Aura version 4.4.0. (I'm not sure if the [documentation](https://neo4j.com/developer/kb/understanding-how-merge-works/#_merge_acquires_locks_on_nodes_and_relationships_in_the_pattern_when_resulting_in_pattern_creation) is still true for 4.3 and above. It seems to stop at 4.2).
- Operating system: Linux Mint 20.1 Cinnamon

**Comment by maiieul** [comment]
After more investigation, I found that the issue actually comes from wanting to increment, then decrement, then increment, etc., through `SET` after a `MATCH` clause.

I found another use case where I need to lock before a match without any `DELETE` clause. To reproduce:

1. create user, post and merge relationship

```
CREATE (p:Post{id:"1", numberOfUpvotes: 0})
CREATE (u:User{id:"2"})
MERGE (u)-[:UPVOTED_POST]->(p)
SET up.atDateTime = [datetime()], p.numberOfUpvotes = p.numberOfUpvotes + 1
RETURN p, u
```

2. Same as above, you can run it concurrently to see it increment or decrement more than once (here, the goal was to decrement p.numberOfUpvotes when the number of elements in the up.atDateTime list is even and increment p.numberOfUpvotes when it's odd).

```
MATCH (p:Post{id:"1"})
MATCH (u:User{id:"2"})
MATCH (u)-[up:UPVOTED_POST]->(p)
CALL apoc.util.sleep(15000)
SET 
    up.atDateTime = up.atDateTime + datetime(),
    p.numberOfUpvotes = CASE WHEN size(up.atDateTime)%2 = 0 THEN p.numberOfUpvotes - 1 ELSE p.numberOfUpvotes + 1 END
RETURN p
```

My workaround:
```
MATCH (p:Post{id:"1"})
MATCH (u:User{id:"2"})
CALL apoc.lock.nodes([p, u])
MATCH (u)-[up:UPVOTED_POST]->(p)
CALL apoc.util.sleep(15000)
SET 
    up.atDateTime = up.atDateTime + datetime(),
    p.numberOfUpvotes = CASE WHEN size(up.atDateTime)%2 = 0 THEN p.numberOfUpvotes - 1 ELSE p.numberOfUpvotes + 1 END
RETURN p
```

Just like `MERGE`, I think that performing another `MATCH` when the `SET` acquires the locks could perhaps solve the issue. It would be nice to at least document this behavior if there are situations where my proposed solution creates other issues.

I can close this issue and reopen another one (maybe as bug?) if it helps for clarity.

**Comment by maiieul** [comment]
Closing, as I opened #12823 for clarity.
